### ISSUE TITLE
Windows service installed in docker container not using Environment Variables

### ISSUE THREAD

**Issue by snowfire20** [aspnet/MicrosoftConfigurationBuilders#25]
I've been trying to convert an asp.net windows service (not .net core) to a docker service. I've successfully installed the service via powershell. when the image is run, the environment variables configured in the app.config are not replaced with the ones supplied in the services app.config.
I've previously tested this service as a console application in the docker container, the variables were correctly replaced.
Is there something special I have to do to have this work for a service?
Or is this not supported?
Thank you

**Comment by StephenMolloy** [comment]
It is my understanding that Windows containers only modify the environment block of the entry point process. Since ASP.Net/IIS is a service, any environment variables set when starting the container wouldn't not be visible to that service.

However, the [ASP.Net](https://github.com/microsoft/aspnet-docker) and [IIS](https://github.com/microsoft/iis-docker) container images rely on a ServiceMonitor.exe utility to be the entry point for the container, and this utility should be proactively modifying the environment of the worker process with any additional environment variables passed to docker run. Is this not the behavior you are seeing?

**Comment by snowfire20** [comment]
This is a separate custom application installed in the image as part of the docker file. it is set to automatic startup. it is not the entry point of the container. right now it trys to run at startup, but can't because the app.config fields have not been replaced with environment variables as expected.

**Comment by StephenMolloy** [comment]
Are you saying that no environment variables get pulled into your config, or just the non-default ones you set when starting the container. If its the former, it would be helpful to see your stripped-down app.config to figure out where things are going wrong. If it's the latter though...

That isn't an issue with the config builder. Non-entrypoint services and processes don't see those environment variables that get set when starting windows containers. As I mentioned before, ASP.Net and IIS containers work around this issue by proactively pushing non-default variables as seen by the entrypoint process into the environment for app pools. [ServiceMonitor.exe](https://github.com/Microsoft/IIS.ServiceMonitor) is the process that does that if you want to see how they handle this situation.

**Comment by snowfire20** [comment]
at the moment I only care about the ones I supply at run time. If this is not supported, can you please update the readme to make that clear. right now all it says is to add these to the app.config and they will work, no limitations are listed in the type of app that it will work with.

**Comment by StephenMolloy** [comment]
As mentioned before, this is not an issue or restriction with the config builders. The config builder can only read the environment variables that exist in the context of it's process. The issue you're running into here is that Windows containers only set those command line environment variables for the entrypoint process, and your app is running as a service instead of the entrypoint.

You could consider two approaches to solving your issue. 1) Run your application as the entrypoint instead of as a service. This is arguably the more container/microservice architecture-friendly way to employ containers. Or 2) do something similar to what I've linked above in IIS containers with ServiceMonitor.exe as the entrypoint to inject those command-line environment variables into your service's environment before starting it.

**Comment by snowfire20** [comment]
1. right now I can't run it as the entry point due to another issue with how the containers work.
2. writing a powershell script right now to replace the variables in the config with env variables.
I still think you should document that services can't use the config builders. your comment about the windows containers not setting them does not appear anywhere in any documentation I could find.

**Comment by StephenMolloy** [comment]
#31

**Comment by pkharris** [comment]
I'm running into the same problem.  Yet when I run as a service in a Linux container... no issue at all.
